mailbox/lucene/src/main/java/org/apache/james/mailbox/lucene/search/LuceneMessageSearchIndex.java

```diff
@@ -431,24 +452,36 @@ public void setEnableSuffixMatch(boolean suffixMatch) {
     public Iterator<MessageUid> search(MailboxSession session, Mailbox mailbox, SearchQuery searchQuery) throws MailboxException {
         Preconditions.checkArgument(session != null, "'session' is mandatory");
         MailboxId mailboxId = mailbox.getMailboxId();
-        Multimap<MailboxId, MessageUid> results = 
-                searchMultimap(
-                    session, 
-                    MultimailboxesSearchQuery
-                        .from(searchQuery)
-                        .inMailboxes(mailboxId)
-                        .build());
-        return results.get(mailboxId).iterator();
+        return FluentIterable.from(searchMultimap(
+            MultimailboxesSearchQuery
+                .from(searchQuery)
+                .inMailboxes(mailboxId)
+                .build()))
+            .transform(new Function<SearchResult, MessageUid>() {
+                @Override
+                public MessageUid apply(SearchResult input) {
+                    return input.getMessageUid();
+                }
+            })
+            .iterator();
     }
 
     @Override
-    public Map<MailboxId, Collection<MessageUid>> search(MailboxSession session, MultimailboxesSearchQuery searchQuery) throws MailboxException {
+    public List<MessageId> search(MailboxSession session, MultimailboxesSearchQuery searchQuery, long limit) throws MailboxException {
         Preconditions.checkArgument(session != null, "'session' is mandatory");
-        return searchMultimap(session, searchQuery).asMap();
+        return FluentIterable.from(searchMultimap(searchQuery))
+            .transform(new Function<SearchResult, MessageId>() {
+                @Override
+                public MessageId apply(SearchResult input) {
+                    return input.getMessageId().get();
+                }
+            })
+            .limit(Long.valueOf(limit).intValue())
+            .toList();
     }
     
-    private Multimap<MailboxId, MessageUid> searchMultimap(MailboxSession session, MultimailboxesSearchQuery searchQuery) throws MailboxException {
-        Multimap<MailboxId, MessageUid> results = LinkedHashMultimap.create();
+    private List<SearchResult> searchMultimap(MultimailboxesSearchQuery searchQuery) throws MailboxException {

```

Comment:
Security is supposed to be checked on Manager layer ?
IMAP search is done on a selected mailbox. If you can select a mailbox, you can search it.
For JMAP, and multiMailbox search, I guess we added security on JMAP layer. Anyway, that is an Open question not really related to this PR, and we can ticket about it.